File: crates/monty-js/src/frame.ts
// Wire framing for the subprocess protocol: each protobuf message is
// prefixed with a 4-byte unsigned little-endian length. Mirrors
// monty-proto's frame.rs, including the 256 MiB cap that stops a corrupted
// or byzantine child from making the parent allocate unbounded memory.

import type { Readable, Writable } from 'node:stream'

/** Maximum accepted frame payload (256 MiB), matching monty-proto. */
export const MAX_FRAME_LEN = 256 * 1024 * 1024

/**
 * Thrown on framing-level corruption (oversized length prefix, EOF mid-frame).
 * Unrecoverable for the connection — the worker must be discarded.
 */
export class FrameError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FrameError'
  }
}

/**
 * Buffered frame reader over a worker's stdout.
 *
 * The protocol is strictly alternating, so a single pending `read()` at a
 * time is all we need: data events accumulate into a buffer and the pending
 * read resolves as soon as a complete frame is available. A clean EOF at a
 * frame boundary resolves to `null`; EOF mid-frame rejects with
 * [`FrameError`] (the worker died while writing).
 */
export class FrameReader {
  private chunks: Buffer[] = []
  private buffered = 0
  private ended = false
  private streamError: Error | null = null
  private pending: { resolve: (frame: Buffer | null) => void; reject: (err: Error) => void } | null = null

  constructor(stream: Readable) {
    stream.on('data', (chunk: Buffer) => {
      this.chunks.push(chunk)
      this.buffered += chunk.length
      this.poll()
    })
    stream.on('end', () => {
      this.ended = true
      this.poll()
    })
    stream.on('error', (err: Error) => {
      this.streamError = err
      this.poll()
    })
  }

  /** Reads the next frame payload, or `null` on clean EOF. */
  read(): Promise<Buffer | null> {
    if (this.pending) {
      return Promise.reject(new FrameError('concurrent frame reads are not supported'))
    }
    return new Promise((resolve, reject) => {
      this.pending = { resolve, reject }
      this.poll()
    })
  }

  /** Resolves the pending read if a full frame (or EOF/error) is available. */
  private poll(): void {
    const pending = this.pending
    if (!pending) {
      return
    }
    if (this.buffered >= 4) {
      const header = this.peek(4)
      const len = header.readUInt32LE(0)
      if (len > MAX_FRAME_LEN) {
        this.pending = null
        pending.reject(new FrameError(`frame of ${len} bytes exceeds maximum of ${MAX_FRAME_LEN}`))
        return
      }
      if (this.buffered >= 4 + len) {
        this.consume(4)
        const frame = this.take(len)
        this.pending = null
        pending.resolve(frame)
        return
      }
    }
    if (this.streamError) {
      this.pending = null
      pending.reject(new FrameError(`error reading from worker: ${this.streamError.message}`))
    } else if (this.ended) {
      this.pending = null
      if (this.buffered === 0) {
        pending.resolve(null)
      } else {
        pending.reject(new FrameError('worker closed mid-frame'))
      }
    }
  }

  /** Returns the first `n` buffered bytes without consuming them. */
  private peek(n: number): Buffer {
    if (this.chunks.length > 0 && this.chunks[0]!.length >= n) {
      return this.chunks[0]!
    }
    // Compact so the header is contiguous; rare in practice.
    const merged = Buffer.concat(this.chunks)
    this.chunks = [merged]
    return merged
  }

  /** Drops `n` buffered bytes. */
  private consume(n: number): void {
    this.take(n)
  }

  /** Removes and returns the first `n` buffered bytes. */
  private take(n: number): Buffer {
    const parts: Buffer[] = []
    let needed = n
    while (needed > 0) {
      const head = this.chunks[0]!
      if (head.length <= needed) {
        parts.push(head)
        needed -= head.length
        this.chunks.shift()
      } else {
        parts.push(head.subarray(0, needed))
        this.chunks[0] = head.subarray(needed)
        needed = 0
      }
    }
    this.buffered -= n
    return parts.length === 1 ? parts[0]! : Buffer.concat(parts)
  }
}

/**
 * Writes one framed message to the worker's stdin, waiting out backpressure
 * so the strict request/event alternation can never deadlock on a full pipe.
 */
export function writeFrame(stream: Writable, payload: Uint8Array): Promise<void> {
  const header = Buffer.allocUnsafe(4)
  header.writeUInt32LE(payload.length, 0)
  const frame = Buffer.concat([header, payload])
  return new Promise((resolve, reject) => {
    stream.write(frame, (err) => (err ? reject(err) : resolve()))
  })
}

File: crates/monty-proto/src/frame.rs
//! Length-prefixed framing for protocol messages over byte streams.
//!
//! Each frame is a 4-byte unsigned **little-endian** length prefix followed by
//! that many bytes of protobuf. LE matches the convention used elsewhere in
//! monty's serialization and is trivial to implement in any language
//! (`readUInt32LE` in Node, `struct.unpack('<I', ...)` in Python).
//!
//! The reader enforces [`MAX_FRAME_LEN`] so a corrupted or byzantine peer
//! cannot make the receiving process allocate unbounded memory from a single
//! bogus length prefix. Writers flush after every frame — the protocol is a
//! strict alternation, so an unflushed frame would deadlock both sides.

use std::{
    error, fmt,
    io::{self, Read, Write},
};

use prost::Message;

/// Default maximum frame length (256 MiB).
///
/// Far above any sane payload, but small enough that a corrupted length
/// prefix cannot trigger a multi-gigabyte allocation in the receiver.
pub const MAX_FRAME_LEN: u32 = 256 * 1024 * 1024;

/// Framing or decoding failure while reading or writing protocol messages.
#[derive(Debug)]
pub enum FrameError {
    /// Underlying stream I/O failure (includes broken pipes — peer death).
    Io(io::Error),
    /// Frame contents were not a valid protobuf message.
    Decode(prost::DecodeError),
    /// Length prefix exceeded the reader's maximum frame length.
    FrameTooLarge {
        /// Length claimed by the prefix.
        len: u32,
        /// The reader's configured maximum.
        max: u32,
    },
    /// The stream ended mid-frame: the peer died while writing.
    Truncated,
}

impl fmt::Display for FrameError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io(e) => write!(f, "frame I/O error: {e}"),
            Self::Decode(e) => write!(f, "frame decode error: {e}"),
            Self::FrameTooLarge { len, max } => write!(f, "frame of {len} bytes exceeds maximum of {max} bytes"),
            Self::Truncated => f.write_str("stream ended mid-frame"),
        }
    }
}

impl error::Error for FrameError {}

impl From<io::Error> for FrameError {
    fn from(e: io::Error) -> Self {
        Self::Io(e)
    }
}

/// Writes length-prefixed protobuf frames to a byte stream.
///
/// Flushes after every frame; see the module docs for why this is required.
#[derive(Debug)]
pub struct FrameWriter<W: Write> {
    inner: W,
}

impl<W: Write> FrameWriter<W> {
    /// Wraps a byte stream. Pass a buffered writer (e.g. `BufWriter<Stdout>`)
    /// when the underlying stream is unbuffered.
    pub fn new(inner: W) -> Self {
        Self { inner }
    }

    /// Direct access to the underlying stream — for tests that need to write
    /// raw (malformed) bytes.
    pub fn get_mut(&mut self) -> &mut W {
        &mut self.inner
    }

    /// Encodes `msg` and writes it as one frame, then flushes.
    pub fn write(&mut self, msg: &impl Message) -> Result<(), FrameError> {
        let len = u32::try_from(msg.encoded_len()).map_err(|_| FrameError::FrameTooLarge {
            len: u32::MAX,
            max: MAX_FRAME_LEN,
        })?;
        // encode_to_vec cannot fail (Vec<u8> grows as needed)
        let body = msg.encode_to_vec();
        self.inner.write_all(&len.to_le_bytes())?;
        self.inner.write_all(&body)?;
        self.inner.flush()?;
        Ok(())
    }
}

/// Reads length-prefixed protobuf frames from a byte stream.
#[derive(Debug)]
pub struct FrameReader<R: Read> {
    inner: R,
    max_frame_len: u32,
}

impl<R: Read> FrameReader<R> {
    /// Wraps a byte stream with the default [`MAX_FRAME_LEN`].
    pub fn new(inner: R) -> Self {
        Self {
            inner,
            max_frame_len: MAX_FRAME_LEN,
        }
    }

    /// Wraps a byte stream with a custom maximum frame length.
    pub fn with_max_frame_len(inner: R, max_frame_len: u32) -> Self {
        Self { inner, max_frame_len }
    }

    /// Reads one frame and decodes it as `M`.
    ///
    /// Returns `Ok(None)` on a clean EOF at a frame boundary (the peer closed
    /// the stream between messages). EOF *inside* a frame is
    /// [`FrameError::Truncated`] — the peer died mid-write.
    pub fn read<M: Message + Default>(&mut self) -> Result<Option<M>, FrameError> {
        let mut len_bytes = [0u8; 4];
        match read_exact_or_eof(&mut self.inner, &mut len_bytes)? {
            ReadOutcome::CleanEof => return Ok(None),
            ReadOutcome::Truncated => return Err(FrameError::Truncated),
            ReadOutcome::Filled => {}
        }
        let len = u32::from_le_bytes(len_bytes);
        if len > self.max_frame_len {
            return Err(FrameError::FrameTooLarge {
                len,
                max: self.max_frame_len,
            });
        }
        let mut body = vec![0u8; len as usize];
        match read_exact_or_eof(&mut self.inner, &mut body)? {
            ReadOutcome::Filled => {}
            // EOF after a length prefix is always mid-frame.
            ReadOutcome::CleanEof | ReadOutcome::Truncated => return Err(FrameError::Truncated),
        }
        M::decode(body.as_slice()).map(Some).map_err(FrameError::Decode)
    }
}

/// Outcome of [`read_exact_or_eof`].
enum ReadOutcome {
    /// The buffer was completely filled.
    Filled,
    /// EOF before any byte was read.
    CleanEof,
    /// EOF after some but not all bytes were read.
    Truncated,
}

/// Like `read_exact` but distinguishes "EOF at the boundary" from "EOF
/// mid-buffer", which the framing layer must report differently.
fn read_exact_or_eof(reader: &mut impl Read, buf: &mut [u8]) -> io::Result<ReadOutcome> {
    let mut filled = 0;
    while filled < buf.len() {
        match reader.read(&mut buf[filled..]) {
            Ok(0) => {
                return Ok(if filled == 0 {
                    ReadOutcome::CleanEof
                } else {
                    ReadOutcome::Truncated
                });
            }
            Ok(n) => filled += n,
            Err(e) if e.kind() == io::ErrorKind::Interrupted => {}
            Err(e) => return Err(e),
        }
    }
    Ok(ReadOutcome::Filled)
}
